Document how to import resources into Terraform

ptgott opened the pull request:
Closes #9720

Add a guide to importing Teleport resources into Terraform. While it is mostly straightforward to add an `import` block, determining the ID to use for each resource requires some planning, since the ID depends on the resource type. 

This change also edits the Terraform provider guide, since testing out this guide was required to test out the resource import workflow:

- Minor style and readability tweaks.
- Remove the `-f` flag in `tctl create`: This way a user doesn't accidentally override an existing `terraform` role
- Add a `tbot` data directory in an unprivileged path in the working directory of the setup script. This way, we don't need to give `teleport` ownership of the default Machine ID output directory.
- Fix identity file paths. These should point to a file called `identity` in the `terraform-identity` directory.
- Edit the "Next steps" section to point to the resource import guide.

ptgott committed:
Document how to import resources into Terraform

Closes #9720

Add a guide to importing Teleport resources into Terraform. While it is
mostly straightforward to add an `import` block, determining the ID to
use for each resource requires some planning, since the ID depends on
the resource type.

This change also edits the Terraform provider guide, since testing out
this guide was required to test out the resource import workflow:

- Minor style and readability tweaks.
- Remove the `-f` flag in `tctl create`: This way a user doesn't
  accidentally override an existing `terraform` role
- Add a `tbot` data directory in an unprivileged path in the working
  directory of the setup script. This way, we don't need to give
  `teleport` ownership of the default Machine ID output directory.
- Fix identity file paths. These should point to a file called
  `identity` in the `terraform-identity` directory.
- Edit the "Next steps" section to point to the resource import guide.

r0mant commented on docs/config.json:
```diff
@@ -644,7 +644,11 @@
             {
               "title": "Terraform Provider",
               "slug": "/management/dynamic-resources/terraform-provider/"
-            }
+            },
+	    {
+	      "title": "Import Teleport Resources into Terraform",
+	      "slug": "/management/dynamic-resources/import-terraform-resources/"
+	    }
```
It's not related to this PR per-se, but I've always found our terraform documentation to be buried and non-discoverable. Whenever I go to search something terraform-provider-related, I always get confused where I should look (is it managing cluster? deploying agents? reference?), and can never find what I'm looking for "organically" and end up just using search.

What do think about elevating all Terraform Provider relevant documentation to its own top-level section in the left nav bar? In fact, I would probably want to do the same for Kube Operator later. I think we had both of them at the top level long time ago but then they got buried during some docs restructuring.

ptgott replied:
In the past, these guides were in `docs/pages/management/guides`, which was kind of a catch-all section with no explicit theme.

In the current docs, we put guides related to dynamic resources in `/management/dynamic-resources`. Reference guides are mostly but not always in "Reference", and guides related to deploying agents are in `/agents`. Since the "Deploy Agents with Terraform" guide has to do with agents, and not dynamic resources, I put it in the `/agents` section.

I can see how this would be confusing if you're looking for all content related to Terraform. In general, the docs are [inconsistent](https://github.com/gravitational/teleport/issues/24120) in where they store reference information, so I'm not sure of a good solution for the reference guide without taking more time to look at this.

On my 4k monitor at the default zoom, the left sidebar already extends past the bottom of the viewport—I'm also not sure if we can accommodate more sidebar sections, but it might be worth looking into. Would it make sense to promote the "Dynamic Resources" docs to a top-level section that includes content for the Kubernetes operator and Terraform provider?

Tener replied:
I can't for the life of me navigate around the docs hierarchy myself. I'll typically just ask AI bot for pointers rather than hunting the menus. I'd really appreciate if you could simply `cmd-f` to find strings in there... but the menu items are folded unless you are visiting a page deeper in the hierarchy.

Tener commented on docs/config.json:
```diff
@@ -644,7 +644,11 @@
-            }
+            },
+	    {
+	      "title": "Import Teleport Resources into Terraform",
+	      "slug": "/management/dynamic-resources/import-terraform-resources/"
```
I like the distinct title of this article, but I'm left puzzled over _Terraform Provider_ in contrast. Might be worth clarifying the goal of that other guide too?

Tener commented on docs/pages/management/dynamic-resources/import-terraform-resources.mdx:
```diff
@@ -0,0 +1,108 @@
+
+## Prerequisites
+
+- The Teleport Terraform provider. Read the [Terraform provider getting started
+  guide](./terraform-provider.mdx) for how to set up Terraform to manage
+  Teleport dynamic resources. This guide assumes that you have set up a root
+  Terraform module for Teleport on your workstation and started `tbot` to to
+  retrieve credentials for the provider.
```
We reference the main guide here as a prerequisite, but it isn't immediately clear why the steps in the main guide are not enough or if only _part_ of those steps should be followed.

ptgott replied:
I've suggested merging the import guide into the setup guide. Since I'm no longer including a comprehensive table of resource IDs, the instructions are a lot shorter, and it's easier to incorporate into the setup guide. This also makes the relationship between these sets of instructions clearer.

Tener replied:
Sounds good, thanks!

Tener commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
```diff
@@ -191,7 +197,12 @@ To apply the configuration:
 
 ## Next steps
 
+- If you have already created Teleport resources with another tool, such as
+  `tctl`, learn how to [import Teleport resources into
+  Terraform](./import-terraform-resources.mdx).
```
The consequences of not doing that may not be immediately clear to the reader.

ptgott replied:
Added this to the intro of the new section I added to the setup guide.

Tener commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
```diff
@@ -121,6 +126,7 @@ paste in the following content:
 
    tctl bots rm terraform-test
    rm -rf ./terraform-identity
+   rm -rf ./tbot-data
```
```suggestion
   rm -rfv ./tbot-data
```

Nit: I try adding verbose to `rm -rf` in case I run it on a wrong thing... this way at least I can see what got removed.

ptgott commented on docs/pages/management/dynamic-resources/import-terraform-resources.mdx:
```diff
@@ -0,0 +1,108 @@
+The Teleport Terraform provider uses two possible methods to
+define the ID of a resource:
+
+- Using another field within the resource, such as `metadata.name` or
+  `metadata.id`
+- Using a constant value such as `cluster_networking_config`
+
+The following table indicates how the Teleport Terraform provider defines the ID
+of each available Teleport resource:
+
+| Resource                              | How to obtain the ID                |
+|---------------------------------------|-------------------------------------|
+| `teleport_app`                        | Use `metadata.name`                 |
+| `teleport_auth_preference`            | Always `auth_preference`            |
+| `teleport_cluster_maintenance_config` | Always `cluster_maintenance_config` |
+| `teleport_cluster_networking_config`  | Always `cluster_networking_config`  |
+| `teleport_database`                   | Use `metadata.name`                 |
+| `teleport_github_connector`           | Use `metadata.name`                 |
+| `teleport_oidc_connector`             | Use `metadata.name`                 |
+| `teleport_saml_connector`             | Use `metadata.name`                 |
+| `teleport_provision_token`            | Use `metadata.id`                   |
+| `teleport_role`                       | Use `metadata.name`                 |
+| `teleport_session_recording_config`   | Always `session_recording_config`   |
+| `teleport_trusted_cluster`            | Use `metadata.name`                 |
+| `teleport_user`                       | Use `metadata.name`                 |
+| `teleport_login_rule`                 | Use `metadata.name`                 |
+| `teleport_device_trust`               | Use `metadata.name`                 |
+| `teleport_okta_import_rule`           | Use `metadata.name`                 |
+| `teleport_access_list`                | Use `metadata.name`                 |
```
Good point. It looks like we either use `metadata.name` or, if there can only be one instance of a resource, the resource name without the `teleport_` prefix. I can remove the table and describe that convention, noting that the only exception is `teleport_provision_token`. How does that sound?

ptgott committed:
Respond to Tener feedback

- Edit the "Terraform Provider" guide title to describe the purpose of
  the guide more explicitly, distinguishing it from the new guide.
- Describe general rules for finding the ID of a resource, rather than
  using a comprehensive table.
- Incorporate the import instructions into the setup guide. This way, we
  don't need to clarify the relationship between two guides, which
  wasn't entirely clear at the outset.
- Add a more explicit benefit statement for importing resources.
- Use the `-v` flag in `rm` commands

ptgott commented:
@Tener @r0mant Friendly ping if you have a moment for a second round of feedback. The change only modifies a single file now (the getting started guide to the Teleport Terraform provider), so while information architecture of our Terraform content is still something we need to improve, I would argue that it's less relevant to this specific PR.

r0mant commented on docs/pages/management/dynamic-resources/terraform-provider.mdx:
```diff
@@ -193,9 +203,88 @@ To apply the configuration:
+   | `teleport_cluster_maintenance_config` | `cluster_maintenance_config` |
+   | `teleport_cluster_networking_config`  | `cluster_networking_config`  |
+
+   For all othe resources, the ID is always the `metadata.name` of the resource.
```
```suggestion
   For all other resources, the ID is always the `metadata.name` of the resource.
```

ptgott committed:
Update docs/pages/management/dynamic-resources/terraform-provider.mdx

Co-authored-by: Roman Tkachenko <roman@goteleport.com>
